File: src/main/java/com/lessmarkup/userinterface/nodehandlers/common/FlatPageNodeHandler.java
package com.lessmarkup.userinterface.nodehandlers.common;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;
import com.lessmarkup.framework.helpers.DependencyResolver;
import com.lessmarkup.framework.helpers.JsonSerializer;
import com.lessmarkup.framework.helpers.StringHelper;
import com.lessmarkup.framework.nodehandlers.AbstractNodeHandler;
import com.lessmarkup.interfaces.cache.DataCache;
import com.lessmarkup.interfaces.security.CurrentUser;
import com.lessmarkup.interfaces.structure.*;
import com.lessmarkup.userinterface.model.common.FlatPagePosition;
import com.lessmarkup.userinterface.model.common.FlatPageSettingsModel;
import com.lessmarkup.userinterface.model.structure.LoadNodeViewModel;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.OptionalLong;

public abstract class FlatPageNodeHandler extends AbstractNodeHandler {

    class FlatNodeEntry {
        private String title;
        private JsonObject viewData;
        private String viewBody;
        private Class<? extends NodeHandler> handlerType;
        private long nodeId;
        private String settings;
        private String anchor;
        private String uniqueId;
        private int level;
        private String path;
        private String fullPath;
        private NodeAccessType accessType;
        private CachedNodeInformation source;

        public void setTitle(String title) {
            this.title = title;
        }
        public String getTitle() {
            return title;
        }

        public void setViewData(JsonObject viewData) {
            this.viewData = viewData;
        }
        public JsonObject getViewData() {
            return viewData;
        }

        public void setViewBody(String viewBody) {
            this.viewBody = viewBody;
        }
        public String getViewBody() {
            return viewBody;
        }

        public void setHandlerType(Class<? extends NodeHandler> handlerType) {
            this.handlerType = handlerType;
        }
        public Class<? extends NodeHandler> getHandlerType() { return handlerType; }

        public void setNodeId(long nodeId) {
            this.nodeId = nodeId;
        }
        public long getNodeId() { return nodeId; }

        public void setSettings(String settings) {
            this.settings = settings;
        }
        public String getSettings() {
            return settings;
        }

        public void setAnchor(String anchor) {
            this.anchor = anchor;
        }
        public String getAnchor() {
            return anchor;
        }

        public void setUniqueId(String uniqueId) {
            this.uniqueId = uniqueId;
        }
        public String getUniqueId() {
            return uniqueId;
        }

        public void setLevel(int level) {
            this.level = level;
        }
        public int getLevel() {
            return level;
        }

        public void setPath(String path) {
            this.path = path;
        }
        public String getPath() {
            return path;
        }

        public void setFullPath(String fullPath) {
            this.fullPath = fullPath;
        }
        public String getFullPath() {
            return fullPath;
        }

        public void setAccessType(NodeAccessType accessType) {
            this.accessType = accessType;
        }
        public NodeAccessType getAccessType() {
            return accessType;
        }

        public void setSource(CachedNodeInformation source) {
            this.source = source;
        }
        public CachedNodeInformation getSource() {
            return source;
        }
    }

    class TreeNodeEntry {
        private String anchor;
        private String title;
        private final List<TreeNodeEntry> children = new ArrayList<>();

        public void setAnchor(String anchor) {
            this.anchor = anchor;
        }
        public String getAnchor() {
            return anchor;
        }

        public void setTitle(String title) {
            this.title = title;
        }
        public String getTitle() {
            return title;
        }

        public List<TreeNodeEntry> getChildren() {
            return children;
        }
    }

    private final List<FlatNodeEntry> flatNodeList = new ArrayList<>();
    private TreeNodeEntry treeRoot;
    private final List<String> scripts = new ArrayList<>();

    private final DataCache dataCache;
    private final CurrentUser currentUser;

    protected FlatPageNodeHandler(DataCache dataCache, CurrentUser currentUser) {
        this.dataCache = dataCache;
        this.currentUser = currentUser;
    }

    private void fillFlatAndTreeList(CachedNodeInformation parent, List<FlatNodeEntry> nodes, TreeNodeEntry parentTreeNode, String anchor, int level, int maxLevel) {
        for (CachedNodeInformation child : parent.getChildren()) {
            if (child.getHandlerType() == null) {
                continue;
            }

            if (child.getHandlerType().equals(FlatPageNodeHandler.class) || !child.isVisible()) {
                continue;
            }

            NodeAccessType accessType = child.checkRights(currentUser);

            if (accessType == NodeAccessType.NO_ACCESS) {
                continue;
            }

            String childAnchor = StringHelper.isNullOrEmpty(anchor) ? "" : anchor + "_";
            FlatNodeEntry entry = new FlatNodeEntry();
            entry.setTitle(child.getTitle());
            entry.setHandlerType(child.getHandlerType());
            entry.setNodeId(child.getNodeId());
            entry.setSettings(child.getSettings());
            entry.setAnchor(childAnchor + child.getPath());
            entry.setLevel(level);
            entry.setPath(child.getFullPath());
            entry.setFullPath(child.getFullPath());
            entry.setSource(child);
            entry.setAccessType(accessType);

            TreeNodeEntry treeNode = new TreeNodeEntry();
            treeNode.setTitle(entry.getTitle());
            treeNode.setAnchor(entry.getAnchor());

            parentTreeNode.getChildren().add(treeNode);

            nodes.add(entry);

            if (level < maxLevel) {
                fillFlatAndTreeList(child, nodes, treeNode, entry.getAnchor(), level + 1, maxLevel);
            }
        }
    }

    @Override
    protected Object initialize() {
        FlatPageSettingsModel settingsModel = getSettings(FlatPageSettingsModel.class);

        NodeCache nodeCache = dataCache.get(NodeCache.class);

        treeRoot = new TreeNodeEntry();

        if (getObjectId().isPresent()) {
            CachedNodeInformation currentNode = nodeCache.getNode(getObjectId().getAsLong());
            fillFlatAndTreeList(currentNode, flatNodeList, treeRoot, "", 1, (settingsModel != null && settingsModel.getLevelToLoad() != 0) ? settingsModel.getLevelToLoad() : 2);
        }

        if (settingsModel == null || settingsModel.isLoadOnShow()) {
            for (FlatNodeEntry node : flatNodeList) {
                NodeHandler handler = createChildHandler(node.getHandlerType());
                JsonElement nodeSettings = null;

                for (String script : handler.getScripts()) {
                    addScript(script);
                }

                for (String stylesheet : handler.getStylesheets()) {
                    addStylesheet(stylesheet);
                }

                if (handler.getSettingsModel() != null && !StringHelper.isNullOrEmpty(node.getSettings())) {
                    nodeSettings = JsonSerializer.deserializeToTree(node.getSettings());
                }

                handler.initialize(OptionalLong.of(node.getNodeId()), 
                        nodeSettings != null && nodeSettings.isJsonObject() ? nodeSettings.getAsJsonObject() : null, 
                        node.getPath(), node.getFullPath(), node.getAccessType());

                node.setViewData(handler.getViewData());
                node.setViewBody(LoadNodeViewModel.getViewTemplate(handler, dataCache));

                List<String> handlerScripts = handler.getScripts();

                if (handlerScripts != null) {
                    handlerScripts.addAll(handlerScripts);
                }
            }

            flatNodeList.removeIf(n -> n.getViewBody() == null);
        }

        int pageIndex = 1;

        for (FlatNodeEntry node : flatNodeList) {
            node.setUniqueId(String.format("flatpage%d", pageIndex++));
        }

        return null;
    }

    @Override
    public JsonObject getViewData() {
        FlatPageSettingsModel settingsModel = getSettings(FlatPageSettingsModel.class);

        JsonObject ret = new JsonObject();
        
        JsonArray children = new JsonArray();
        for (TreeNodeEntry child : treeRoot.getChildren()) {
            children.add(JsonSerializer.serializePojoToTree(child));
        }

        ret.add("tree", children);

        JsonArray flatArray = new JsonArray();
        for (FlatNodeEntry entry : flatNodeList) {
            JsonObject o = new JsonObject();
            o.addProperty("anchor", entry.getAnchor());
            o.addProperty("level", entry.getLevel());
            o.addProperty("nodeId", entry.getNodeId());
            o.addProperty("path", entry.getPath());
            o.addProperty("title", entry.getTitle());
            o.addProperty("uniqueId", entry.getUniqueId());
            o.addProperty("viewBody", entry.getViewBody());
            o.add("viewData", entry.getViewData());
            flatArray.add(o);
        }

        ret.add("flat", flatArray);
        ret.addProperty("position", settingsModel != null ? settingsModel.getPosition().toString() : FlatPagePosition.RIGHT.toString());
        JsonArray scriptsArray = new JsonArray();
        scripts.forEach(s -> scriptsArray.add(new JsonPrimitive(s)));
        ret.add("scripts", scriptsArray);

        return ret;
    }

    @Override
    public Class<?> getSettingsModel() {
        return FlatPageSettingsModel.class;
    }

    @Override
    public boolean hasChildren() {
        return true;
    }

    private NodeHandler constructHandler(FlatNodeEntry node) {
        NodeAccessType accessType = node.getSource().checkRights(currentUser);

        if (accessType == NodeAccessType.NO_ACCESS) {
            return null;
        }

        NodeHandler handler = DependencyResolver.resolve(node.getHandlerType());

        JsonElement nodeSettings = null;

        if (handler.getSettingsModel() != null && !StringHelper.isNullOrEmpty(node.getSettings())) {
            nodeSettings = JsonSerializer.deserializeToTree(node.getSettings());
        }

        handler.initialize(OptionalLong.of(node.getNodeId()), 
                nodeSettings != null && nodeSettings.isJsonObject() ? nodeSettings.getAsJsonObject() : null,
                node.getPath(), node.getFullPath(), accessType);

        return handler;
    }

    @Override
    public ChildHandlerSettings getChildHandler(String path) {
        Optional<FlatNodeEntry> node = flatNodeList.stream().filter(n -> n.getPath().equals(path)).findFirst();

        if (!node.isPresent()) {
            return null;
        }

        NodeHandler handler = constructHandler(node.get());

        ChildHandlerSettings settings = new ChildHandlerSettings();
        settings.setHandler(handler);
        settings.setId(OptionalLong.of(node.get().getNodeId()));
        settings.setPath(path);
        settings.setTitle(node.get().getTitle());

        return settings;
    }

    @Override
    public Tuple<Object, Method> getActionHandler(String name, JsonObject data) {
        JsonElement nodeIdElement = data.get("flatNodeId");
        if (nodeIdElement != null) {
            long nodeId = nodeIdElement.getAsLong();
            Optional<FlatNodeEntry> node = flatNodeList.stream().filter(n -> n.getNodeId() == nodeId).findFirst();
            if (node.isPresent()) {
                NodeHandler handler = constructHandler(node.get());
                return handler.getActionHandler(name, data);
            }
        }

        return super.getActionHandler(name, data);
    }
}

File: src/main/scala/com/lessmarkup/userinterface/nodehandlers/common/FlatPageNodeHandler.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.userinterface.nodehandlers.common

import java.lang.reflect.Method

import com.google.gson._
import com.lessmarkup.framework.helpers.{JsonSerializer, StringHelper}
import com.lessmarkup.framework.nodehandlers.{AbstractNodeHandler, NodeHandlerConfiguration}
import com.lessmarkup.framework.system.RequestContextHolder
import com.lessmarkup.interfaces.annotations.NodeAccessType
import com.lessmarkup.interfaces.cache.DataCache
import com.lessmarkup.interfaces.structure._
import com.lessmarkup.userinterface.model.common.{FlatPagePosition, FlatPageSettingsModel}
import com.lessmarkup.userinterface.model.structure.LoadNodeViewModel

abstract class FlatPageNodeHandler(dataCache: DataCache, configuration: NodeHandlerConfiguration) extends AbstractNodeHandler(configuration) {

  private class FlatNodeEntry(val title: String,
                              val viewData: Option[JsonObject],
                              val viewBody: String,
                              val handlerType: Class[_ <: NodeHandlerFactory],
                              val nodeId: Long,
                              val settings: String,
                              val anchor: String,
                              val uniqueId: String,
                              val level: Int,
                              val path: String,
                              val fullPath: String,
                              val accessType: NodeAccessType,
                              val source: CachedNodeInformation,
                              val handler: NodeHandler)

  private class TreeNodeEntry(val anchor: String, val title: String, val flatEntry: FlatNodeEntry, val children: Seq[TreeNodeEntry])

  private val treeRoot: Option[TreeNodeEntry] = createRootNode
  private val flatNodeList: Seq[FlatNodeEntry] = if (treeRoot.isDefined) flattenTree(treeRoot.get) else Seq()
  private val scripts: Seq[String] = flatNodeList.flatMap(n => n.handler.getScripts)
  private val stylesheets: Seq[String] = flatNodeList.flatMap(n => n.handler.getStylesheets)

  private def flattenTree(treeNodeEntry: TreeNodeEntry): Seq[FlatNodeEntry] = {
    treeNodeEntry.children.flatMap(flattenTree) :+ treeNodeEntry.flatEntry
  }

  private def createRootNode: Option[TreeNodeEntry] = {
    if (getObjectId.isEmpty) {
      return None
    }

    val nodeCache: NodeCache = dataCache.get(classOf[NodeCache])

    val node = nodeCache.getNode(getObjectId.get)

    if (node.isEmpty) {
      return None
    }

    val settingsModel: Option[FlatPageSettingsModel] = getSettings(classOf[FlatPageSettingsModel])

    val levelToLoad = if (settingsModel != null && settingsModel.get.levelToLoad != 0) settingsModel.get.levelToLoad else 2

    createNode(node.get, "", 0, levelToLoad)
  }

  private def createNode(source: CachedNodeInformation, anchor: String, level: Int, maxLevel: Int): Option[TreeNodeEntry] = {

    val accessType = source.checkRights(RequestContextHolder.getContext.getCurrentUser)

    if (accessType == NodeAccessType.NO_ACCESS) {
      return None
    }

    val nodeSettings: Option[JsonObject] = if (!StringHelper.isNullOrEmpty(source.settings)) {
      Option(JsonSerializer.deserializeToTree(source.settings).getAsJsonObject)
    } else {
      None
    }

    val childHandlerSettings = new NodeHandlerConfiguration(Option(source.nodeId), nodeSettings, accessType, source.path, source.fullPath)
    val childHandler = createChildHandler(source.handlerType.get, childHandlerSettings)

    val viewBody = LoadNodeViewModel.getViewTemplate(childHandler, dataCache)

    if (viewBody == null) {
      return None
    }

    val entry: FlatNodeEntry = new FlatNodeEntry(
      title = source.title,
      viewData = childHandler.getViewData,
      viewBody = viewBody,
      handlerType = source.handlerType.get,
      nodeId = source.nodeId,
      settings = source.settings,
      anchor = anchor,
      uniqueId = s"flatpage${source.nodeId}",
      level = level,
      path = source.fullPath,
      fullPath = source.fullPath,
      accessType = accessType,
      source = source,
      handler = childHandler
    )

    val children = if (level > maxLevel) Seq() else source.children
      .filter(c => c.handlerType != null && c.handlerType.get != classOf[FlatPageNodeHandler] && c.enabled)
      .flatMap(c => {
      val childAnchor = if (StringHelper.isNullOrEmpty(anchor)) "" else anchor + "_"
      createNode(c, childAnchor, level+1, maxLevel)
    })

    Option(new TreeNodeEntry(anchor, source.title, entry, children))
  }

  override def getViewData: Option[JsonObject] = {

    val settingsModel = getSettings(classOf[FlatPageSettingsModel])
    val ret = new JsonObject
    val children = new JsonArray
    if (treeRoot.isDefined) {
      for (child <- treeRoot.get.children) {
        children.add(JsonSerializer.serializePojoToTree(child))
      }
    }
    ret.add("tree", children)
    val flatArray: JsonArray = new JsonArray
    for (entry <- flatNodeList) {
      val o = new JsonObject
      o.addProperty("anchor", entry.anchor)
      o.addProperty("level", entry.level)
      o.addProperty("nodeId", entry.nodeId)
      o.addProperty("path", entry.path)
      o.addProperty("title", entry.title)
      o.addProperty("uniqueId", entry.uniqueId)
      o.addProperty("viewBody", entry.viewBody)
      if (entry.viewData.isDefined) {
        o.add("viewData", entry.viewData.get)
      } else {
        o.add("viewData", JsonNull.INSTANCE)
      }
      flatArray.add(o)
    }
    ret.add("flat", flatArray)
    ret.addProperty("position", if (settingsModel.isDefined) settingsModel.get.position.toString else FlatPagePosition.RIGHT.toString)
    val scriptsArray: JsonArray = new JsonArray
    for(s <- scripts) {
      scriptsArray.add(new JsonPrimitive(s))
    }
    ret.add("scripts", scriptsArray)
    val stylesheetsArray = new JsonArray
    for (s <- stylesheets) {
      stylesheetsArray.add(new JsonPrimitive(s))
    }
    ret.add("stylesheets", stylesheetsArray)
    Option(ret)
  }

  override def getSettingsModel: Option[Class[_]] = {
    Option(classOf[FlatPageSettingsModel])
  }

  override def hasChildren: Boolean = {
    true
  }

  private def constructHandler(node: FlatNodeEntry): NodeHandler = {
    val accessType: NodeAccessType = node.source.checkRights(RequestContextHolder.getContext.getCurrentUser)
    if (accessType eq com.lessmarkup.interfaces.annotations.NodeAccessType.NO_ACCESS) {
      return null
    }

    val nodeSettings =
      if (!StringHelper.isNullOrEmpty(node.settings)) {
        Option(JsonSerializer.deserializeToTree(node.settings).getAsJsonObject)
      } else {
        None
      }

    val configuration = new NodeHandlerConfiguration(Option(node.nodeId), nodeSettings, node.accessType, node.path, node.fullPath)

    createChildHandler(node.handlerType, configuration)
  }

  def getChildHandler(path: String): ChildHandlerSettings = {
    val node = flatNodeList.find(n => n.path == path)
    if (node.isEmpty) {
      return null
    }
    val handler: NodeHandler = constructHandler(node.get)
    new ChildHandlerSettings(handler, Option(node.get.nodeId), node.get.title, node.get.path, Seq())
  }

  override def getActionHandler(name: String, data: JsonObject): Option[(AnyRef, Method)] = {
    val nodeIdElement: JsonElement = data.get("flatNodeId")
    if (nodeIdElement != null) {
      val nodeId: Long = nodeIdElement.getAsLong
      val node = flatNodeList.find(n => n.nodeId == nodeId)
      if (node.isDefined) {
        val handler: NodeHandler = constructHandler(node.get)
        return handler.getActionHandler(name, data)
      }
    }
    super.getActionHandler(name, data)
  }
}